src/libclipper/src/containers.cpp

```diff
@@ -30,78 +34,191 @@ ModelContainer::ModelContainer(VersionedModelId model, int container_id,
+  } else {
+    curr_batch_size = estimate(budget);
   }
-  return batch_size;
+  max_batch_size_ = std::max(curr_batch_size, max_batch_size_);
+  return curr_batch_size;
+}
+
+void ModelContainer::fit_estimator() {
+  std::unique_lock<std::mutex> estimator_lock(estimator_mtx_);
+  boost::shared_lock<boost::shared_mutex> datapoints_lock(datapoints_mtx_);
+
+  size_t num_datapoints = processing_datapoints_.size();
+
+  std::vector<EstimatorLatency> x_vals;
+  x_vals.reserve(num_datapoints);
+  std::vector<EstimatorBatchSize> y_vals;
+  y_vals.reserve(num_datapoints);
+
+  // Calculate the pooled processing latency
+  // variance across all batch sizes. This will
+  // be used to obtain an estimate for the p99
+  // processing latency at each batch size
+  //
+  // Pooled variance (https://en.wikipedia.org/wiki/Pooled_variance)
+  // is used under the empirically tested assumption that latency variance
+  // does not change significantly with batch size
+
+  double pooled_std_num = 0;
+  double pooled_std_denom = -1 * static_cast<double>(num_datapoints);
+  double num_valid_stds = 0;
+
+  for (auto &entry : processing_datapoints_) {
+    LatencyInfo &latency_info = entry.second;
+    double info_size = std::get<0>(latency_info);
+    if (info_size >= MINIMUM_BATCH_SAMPLE_SIZE) {
+      double lats_std = std::get<2>(latency_info);
+      double lats_var = std::pow(lats_std, 2);
+      pooled_std_num += (info_size - 1) * lats_var;
+      pooled_std_denom += info_size;
+    }
+  }
+
+  double pooled_std =
+      std::sqrt(pooled_std_num / std::max(1.0, pooled_std_denom));
+
+  // Using the pooled variance, obtain the
+  // estimated p99 latency for each batch size
+
+  EstimatorLatency fitting_lat;
+  for (auto &entry : processing_datapoints_) {
+    EstimatorBatchSize batch_size = entry.first;
+    LatencyInfo &latency_info = entry.second;
+    double lats_mean = std::get<1>(latency_info);
+    double upper_bound_lat = lats_mean + (LATENCY_Z_SCORE * pooled_std);
+    // Scale the upper bound latency in order to moderate the regularization
+    // term associated with ridge regression
+    fitting_lat(0) = upper_bound_lat * REGRESSION_DATA_SCALE_FACTOR;
+    x_vals.push_back(fitting_lat);
+    y_vals.push_back(std::move(batch_size));
+  }
+
+  datapoints_lock.unlock();
+
+  estimator_ = estimator_trainer_.train(x_vals, y_vals);
+}
+
+size_t ModelContainer::explore() {
+  auto mb_search =
+      processing_datapoints_.find(static_cast<double>(max_batch_size_));
+  if (mb_search != processing_datapoints_.end() &&
+      std::get<0>(mb_search->second) < MINIMUM_BATCH_SAMPLE_SIZE) {
+    // We don't have a large enough latency sample
+    // corresponding to the maximum batch size, so
+    // we won't update the maximum size
+    return max_batch_size_;
+  } else if (max_batch_size_ < ADDITIVE_EXPANSION_THRESHOLD) {
+    return max_batch_size_ + 1;
+  } else {
+    double expansion_factor = exploration_distribution_(exploration_engine_);
```

Comment:
Make sure that `expansion_factor` is non-negative (you can do a `max(0, expansion_factor)`).